### DISCUSSION TITLE
get_node(NodeId) gives not back the same Node depending on how the NodeId was specified

### DISCUSSION THREAD

**Discussion by SnoopAir** [FreeOpcUa/opcua-asyncio#1636] [Q&A]
I'm pretty confused about the the result of get_Node() or better why it matters or better differs how the NodeId was specified.

In the following I create the NodeId once with numeric values for id and namespace only and the other time I specify the identifier as string. 

- Both return a NodeId object. 
- Both can be handed over to get_Node() 
- Both times it will give back a Node object
- BUT: Once it is the correct node object, the other time it's something ... completely rubbish
- At least it realizes, that the 2 node objects are not the same node

My questions are: 

1. Why is the 2nd way not retrieving a correct Node?
2. If this is wrong to do it like this, why does it not throw an exception somewhere but silently fails later one when you try to use the node i.e. to get child nodes.

`

    tmpNodeId_ok = asyncua.ua.NodeId(Identifier=7021, NamespaceIndex=6)
    tmpNodeId_nok = asyncua.ua.NodeId(Identifier="SerialNumber", NamespaceIndex=6)

    logger.info("int: tmpNodeId_ok  '{}'".format(tmpNodeId_ok))
    logger.info("str: tmpNodeId_nok '{}'".format(tmpNodeId_nok))

    tmpNode_ok = server.get_node(tmpNodeId_ok)
    tmpNode_nok = server.get_node(tmpNodeId_nok)

    logger.info("int: tmpNode_ok  '{}'".format(tmpNode_ok))
    logger.info("str: tmpNode_nok '{}'".format(tmpNode_nok))

    if tmpNode_ok == tmpNodeId_nok:
        logger.info("It even pretends to be the same node")
    else:
        logger.info("At least it recognizes it's not the same node!")`

**Comment by schroeder-** [comment]
This is by design of opc ua. You can not check nodeid before, because the addressspace is dynamic. Also a node can only have one nodeid. So a string nodeid is never the same as a numeric nodeid.

**Reply by SnoopAir** [reply]
While I understand, that from an OO point these nodeids might be different, they could still address the same node. I almost wanted to say that they HAVE to point to the same node. But I realized, that I can (and actually do) have nodes in the same namespace with the same browesename.

What I do NOT understand though is, why addressing a nodedi which does not exist at all (in my case using namespace and browsename) gives back a node instead of an error. I only realized that it is a "fake node" as I did not see expected child nodes under it.
Thus: Why do I get back a node instead of an error?

**Reply by schroeder-** [reply]
Browsename != NodeId . If you want to use a browsename, there are other methods, to get the node. Such as ` root.get_child(["4:Folder", "3:MyFolder", ...])` . This operation fails if the node doesn't exist.

There is no way in opc ua to check if a node exists without executing a request. So the api was chosen to give you a node and fail if in the request. Also in a late time in the future the node could also be delete or added from the server. And there are even OPC UA Servers, that have no addresspace, but still allow read and write operations of nodeids (old Siemens PLCs).

If you want to check if a node exists, use `await node.read_browse_name()`

**Reply by SnoopAir** [reply]
Thank you @schroeder- for the fast replies. I'm aware, that browsename != nodeid. But a nodeid can be created using a browsename using string notation as i.e. "ns=2;s=MyTemperature" as mentioned on the UnifiedAutomation website (which is easier to read then the official standard): https://documentation.unified-automation.com/uasdkhp/1.4.1/html/_l2_ua_node_ids.html.
Either this is wrong or (which I don't exclude) I misunderstand the string notation.

IMHO the concept of having "get_child" and "get_node" with different ways to specify a node are rather confusing - at least to me.

But again, and maybe to clarify: I'm NOT talking about the client side, I'm looking at it from the server point of view. So that one needs to know if a node exists and could therefor fail immediately when calling get_node(NODEID_WHICH_DOES_NOT_EXIST).

**Reply by schroeder-** [reply]
get_child -> Uses Browsename / Browsepath to translate to a nodeid, this is done by the server. (Network roundtrip)
get_node -> returns an object to access opc ua services (write, read, browse, ....) via nodeid. This is done on the client side without verification, this is for perfomance reasons, because no  network roundtrip needed.

On the server side you could access the addresspace direct:
```
nodeid = asyncua.ua.NodeId(Identifier=7021, NamespaceIndex=6)
try:
     server.iserver.aspace[nodeid]
except KeyError:
     print(f"{nodeid} not found")
```

**Reply by SnoopAir** [reply]
@schroeder- the latter additional path you showed using simple the server list internally works.

What I still try to figure out is how to use the approach of "get_child" and providing a path. Given the following tree structure, how can I address the node _SerialNumber_  with path?

Objects -> DeviceSets -> ADEVICE -> Identification -> SerialNumber

Identification: namespaceId=6, browseName = "Identification"
SerialNumber: namespaceId=6, browseName = "SerialNumber"

Tested the following

`aDevice.get_child("6:7016")  # -> does NOT work`
`aDevice.get_child("6:Identification") # -> works`
`aDevice.get_child("ns=6:s=Identification") # -> does NOT work`
`aDevice.get_child("ns=6:i=7016") # -> does NOT work`

After that build up on the one which was working providing it as a list (iterable)

`aDevice.get_child(["6:Identification", ]) # -> works also`

But now getting brave and trying to access the leave "SerialNumber" underneath with this approach:

`aDevice.get_child(["6:Identification", "6:SerialNumber"])  # -> DOES NOT WORK`

`aDevice.get_child(["6:Identification", "2:SerialNumber"])  # -> works`

And that's what leaves me puzzled. The namespaceID of both nodes is 6! But in UAExpert I see as well a "2" in front of the browseName, but not really as part of the string, and I have no glue where it is coming from or what's it's meaning.

![2024-05-15_09-30-41](https://github.com/FreeOpcUa/opcua-asyncio/assets/130572255/0f805226-b770-41a7-85ff-7fa1b6d3ef6b)


But it seems obvious from the tests above, that the notation: "2:SerialNumber" is NOT a qualified name in the sense of "namespaceId:name".

@schroeder- could you point me in the direction of documentation clarifying this? Or give me some explanations what I'm not understanding correctly?







Background:

* I'm working on the server side
* Serer will be only a "mock" to test/develop against as long as no real HW is available
* Most of the address space tree structure is already added when adding a device node "ADEVICE"
* Looking for a convenient way to handle access to all this nodes underneath such as setting up default values

**Comment by schroeder-** [comment]
As a always NodeId != BrowseName.
Also a NodeId and BrowseName can have different namespaces and is common.
Also `get_child `only works with BrowseNames or a list of BrowseNames or a RelativePath see [ref](https://reference.opcfoundation.org/Core/Part4/v104/docs/A.2#TableA.2)

RelativePath would be:
`get_child("6:Identification/2:SerialNumber")`

Also you can use QualifiedName instead of a string:
`get_node([ua.QualifiedName(6, "Identification"), ua.QualifiedName(2, "SerialNumber")])`

"2:SerialNumber" is the string format of a QualifiedName.
